lib/spack/spack/concretize.py

```diff
@@ -347,26 +347,46 @@ def concretize_compiler_flags(self, spec):
             # running.
             return True
 
-        compiler_match = lambda other: (
-            spec.compiler == other.compiler and
-            spec.architecture == other.architecture)
+        def compiler_match(other, strict=True):
+            if strict:
+                # Other compiler already matches this compiler.
+                return (spec.compiler == other.compiler and
+                        spec.architecture == other.architecture)
+            else:
+                # Other compiler may match this compiler.
+                if other.compiler and other.compiler.concrete:
+                    return (spec.compiler == other.compiler and
+                            spec.architecture == other.architecture)
+                return True
 
         ret = False
         for flag in spack.spec.FlagMap.valid_compiler_flags():
             if flag not in spec.compiler_flags:
-                spec.compiler_flags[flag] = list()
```

Comment:
Overall my understanding is that the original complaint on the slack channel had to do with flags being added to already-installed packages; the initial check for `spec._concrete` resolves this and the additional changes in this function are not necessary to deal with that problem. That being said my understanding is that this logic is intended to have the following effects:

* Ensure that you do not merge flags from different compilers
* Ensure that when a user explicitly sets compiler flags on a package, that flags from other compilers will not add to it

IMO this could benefit from an added unit test.